## DNA-Seq WXS somatic variant calling

This task takes the "harmonized" WXS BAM files which are the outputs from the previous run [DNA-Seq WXS](../../tasks/WXS/README.md).

It should have 9 pairs with:
* Normal: `C836.MJ.2.gdc_realn.bam`
* Tumor:
  * `C836.MKN74.2.gdc_realn.bam`
  * `C836.SNU-1105.1.gdc_realn.bam`
  * `C836.SNU-1066.4.gdc_realn.bam`
  * `C836.FTC-238.1.gdc_realn.bam`
  * `C836.Hs_766T.2.gdc_realn.bam`
  * `C836.PK-1.1.gdc_realn.bam`
  * `C836.HCC1395.2.gdc_realn.bam`
  * `C836.DV-90.1.gdc_realn.bam`
  * `C836.MDA-MB-415.1.gdc_realn.bam`

The example CWL input json is [here](wxs.variant-calling.input.json)

For the input data, please refer to [input_mapping.json](../../input_mapping/input_mapping.json):
* Reference files: `input_mapping["reference_files"]["WXS somatic variant calling"]`
* Input data: `From previous DNA-Seq WXS alignment.`
  * Please note that the Sanger workflow will need the `.bai` index files generated by the DNA-Seq workflow as well.